File: targets/js/sample-counter-v4/src/components/counter-app.ts
import { Counter } from "#/models";
import { App, StatefulWidget, button, column, heading, row, type BuildContext, type Widget } from "#/mvu";

export class CounterApp extends StatefulWidget<Counter> {
  constructor() {
    super();
  }

  initial(): Counter {
    return Counter.zero;
  }

  protected build(ctx: BuildContext<Counter>): Widget {
    return column({
      gap: 12,
      children: [heading({ content: `Count: ${ctx.state.count}` }), row({
        gap: 8,
        children: [button({
          label: "➖",
          onPressed: () => ctx.setState((s: Counter) => s.decrement()),
        }), button({
          label: "➕",
          onPressed: () => ctx.setState((s: Counter) => s.increment()),
        }), button({
          label: "Reset",
          onPressed: () => ctx.setState((_: Counter) => Counter.zero),
        })],
      })],
    });
  }

  static mount(containerId: string): void {
    App.run(containerId, new CounterApp());
  }
}

File: samples/SampleCounterV3/Components/CounterApp.cs
using SampleCounterV3.Models;
using SampleCounterV3.Mvu;
using static SampleCounterV3.Mvu.Ui;

namespace SampleCounterV3.Components;

public sealed partial class CounterApp : StatefulWidget<Counter>
{
    public override Counter Initial() => Counter.Zero;

    protected override Widget Build(BuildContext<Counter> ctx) =>
        Column(
            gap: 12,
            Heading($"Count: {ctx.State.Count}", level: 1),
            Row(
                gap: 8,
                Button("➖", onPressed: () => ctx.SetState(s => s.Decrement())),
                Button("➕", onPressed: () => ctx.SetState(s => s.Increment())),
                Button("Reset", onPressed: () => ctx.SetState(_ => Counter.Zero))
            )
        );
}
